jax_md/elasticity.py

```diff
@@ -0,0 +1,526 @@
+
+  Mandel notation is a way to represent symmetric second-rank tensors and
+  fourth-rank tensors with minor symmetries in a reduced form. Pairs of indices
+  are combined as follows:
+      
+      For tensors of shape (2,2) or (2,2,2,2):
+      tensor indices        Mandel indices
+      0,0 -------------->   0
+      1,1 -------------->   1
+      0,1 or 1,0 ------->   2
+
+      For tensors of shape (3,3) or (3,3,3,3):
+      tensor indices        Mandel indices
+      0,0 -------------->   0
+      1,1 -------------->   1
+      2,2 -------------->   2
+      1,2 or 2,1 ------->   3
+      0,2 or 2,0 ------->   4
+      0,1 or 1,0 ------->   5
+  
+  If mandel_index(i,j) performs the above index mapping, then the input T and 
+  output M satisfy
+    M[mandel_index(i,j)] = T[i,j] * w
+  or
+    M[mandel_index(i,j), mandel_index(k,l)] = T[i,j,k,l] * w(i,j) * w(k,l)
+  where
+    w(i,j) = 1       if i==j
+           = sqrt(2) if i!=j 
+  is a weight that is used to ensure proper contraction rules. Here (and only 
+  here) we do not assume major symmetries in fourth-rank tensors.
+
+  Args:
+    T: Array with 4 possible shapes:
+      1. T.shape == (2,2)
+         Convert a symmetric array of shape (2,2) to an array of shape (3,)
+      2. T.shape == (3,3)
+         Convert a symmetric array of shape (3,3) to an array of shape (6,)
+      3. T.shape == (2,2,2,2)
+         Convert a tensor of shape (2,2,2,2) with minor symmetries to an array
+         of shape (3,3)
+      4. T.shape == (3,3,3,3)
+         Convert a tensor of shape (3,3,3,3) with minor symmetries to an array
+         of shape (6,6)
+  
+  Output: Array of shape (3,), (6,), (3,3), or (6,6)
+
+  see: https://sbrisard.github.io/janus/mandel.html (accessed 21 April, 2021)
+  """
+  DIM = T.shape[0]
+  if not (DIM==2 or DIM==3):
+    raise AssertionError('DIM must be 2 or 3')
+  
+  rank = len(T.shape)
+  if not (rank==2 or rank==4):
+    raise AssertionError('T must have rank 2 or 4')
+  
+  m_map, weight = _get_mandel_mapping_weight(DIM, T.dtype)
+  
+  if rank == 2:
+    extract = lambda idx, w: T[idx[0], idx[1]] * w
+    M = vmap(extract, in_axes=(0,0))(m_map, weight)
+  else:
+    extract = lambda idx0, idx1, w0, w1: T[idx0[0], idx0[1], idx1[0], idx1[1]] * w0 * w1
+    M = vmap(vmap(extract, in_axes=(0,None,0,None)), in_axes=(None,0,None,0))(m_map, m_map, weight, weight)
+  return M
+
+def mandel_to_tensor(M: Array) -> Array:
+  """ Perform the inverse of M = tensor_to_mandel(T).
+
+  Args:
+    M: Array of shape (3,), (6,), (3,3), or (6,6)
+  
+  Output: Array of shape (2,2), (3,3), (2,2,2,2), or (3,3,3,3)
+  """
+  DIM = M.shape[0]
+  if not (DIM==3 or DIM==6):
+    raise AssertionError('DIM must be 3 or 6')
+  
+  rank = len(M.shape)
+  if not (rank==1 or rank==2):
+    raise AssertionError('T must have rank 1 or 2')
+
+  def mandel_index(i,j):
+    return lax.cond(i==j, lambda ij: ij[0], lambda ij: DIM-ij[0]-ij[1], (i,j))
+
+  if DIM == 3:
```

Comment:
Can you explain this line? Why is dimension 2 if DIM is 3 and vice-versa?

Reply:
Yes! the variable `DIM` isn't strictly the spatial dimension, its the elements of shape of the input. When converting mandel notation to tensor notation, the mandel matrix has shape (3,3) for a 2d system and shape (6,6) for a 3d system. So the `else` on this line could be `elif DIM == 6:`, but this is already checked for above.

Perhaps what this means is that, regarding your comment above about changing `DIM` to `dim`, perhaps we should just call it something else entirely. Thoughts on what to call it? `tensor_dimension`?